### PR TITLE
[MM-52924] Add `ConfigurationWillBeSaved` hook example

### PR DESCRIPTION
**Pull request by streamer45**
#### Summary

Adding an example for the new `ConfigurationWillBeSaved` hook. 

Still not entirely sure how we deal with new hooks on this side given the server dependency is rather out of date. Are we expecting users to manually update it? Otherwise I don't see how new hooks would ever work given the plugin gets compiled without the required changes.

#### Server PR

https://github.com/mattermost/mattermost/pull/23567

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-52924

### CODE REVIEW

File: server/configuration.go [hunk 1]
```diff
@@ -231,6 +231,84 @@ func (p *Plugin) OnConfigurationChange() error {
...
+	js, err := json.Marshal(configData)
+	if err != nil {
+		p.API.LogError(
+			"Failed to marshal config data ConfigurationWillBeSaved",
+			"error", err.Error(),
+		)
+		return nil, nil
+	}
+
+	if err := json.Unmarshal(js, &cfg); err != nil {
```
Comment (hanzei) [inline server/configuration.go:273]: Nit pick: Is is worth allocating another variable here to make it easier to compare new and old config?

**Review by hanzei — APPROVED** [review]
LGTM :+1:

Thanks for taking care of the example, Claudio

File: server/configuration.go (hunk 1 above)
Comment (streamer45) [reply server/configuration.go:273]: Sure, although the example doesn't really do any comparison with the previous one.

**Review by cpoile — APPROVED** [review]
> Still not entirely sure how we deal with new hooks on this side given the server dependency is rather out of date. Are we expecting users to manually update it? Otherwise I don't see how new hooks would ever work given the plugin gets compiled without the required changes.

I'm not really sure what you mean by this, can you reword it?

**Comment by streamer45** [comment]
> > Still not entirely sure how we deal with new hooks on this side given the server dependency is rather out of date. Are we expecting users to manually update it? Otherwise I don't see how new hooks would ever work given the plugin gets compiled without the required changes.
> 
> I'm not really sure what you mean by this, can you reword it?

@cpoile As is if you compile the demo plugin it won't run the hook because it's pointing to a very out of date server version so there isn't much to demo, code is mostly documentation at this point. We need to update everything, which also means the `plugin-api` dependency so not straighforward.

**Comment by cpoile** [comment]
> > > Still not entirely sure how we deal with new hooks on this side given the server dependency is rather out of date. Are we expecting users to manually update it? Otherwise I don't see how new hooks would ever work given the plugin gets compiled without the required changes.
> > 
> > 
> > I'm not really sure what you mean by this, can you reword it?
> 
> @cpoile As is if you compile the demo plugin it won't run the hook because it's pointing to a very out of date server version so there isn't much to demo, code is mostly documentation at this point. We need to update everything, which also means the `plugin-api` dependency so not straighforward.

Ah, ok, gotcha. Have we decided what we're doing with plugin-api @hanzei ? Would it be worth upgrading it, and the demo? (this would be a good chance)

**Pull request merged**